### PR TITLE
Updates to GPS handling

### PR DESCRIPTION
**Pull request by Deatojef**
<ol>
<li>Processes now startup without blocking or waiting on the GPS to obtain a 3D fix.</li>
<li>The APRS-IS connector will not beacon to the APRS-IS cloud (i.e. APRS beaconing over the internet) without a local GPS location fix.  This is similar behavior to Dire Wolf were it configured to beacon over RF or Inet.</li>
<li>Significant improvements to the gpspoller.py process:</li>
<ol>
<li>No longer block waiting on data back from GPSD.  It checks if data is available first. </li>
<li>The /eosstracker/www/gpsstatus.json file is now updated with a list of heard satellites regardless of the GPS fix status (ex. No fix, 2D fix, 3D fix).  The EOSS Tracker web home page will display these in the GPS Status box. </li>
<li>An “error” condition text string now accompanies the GPS status and is included in the gpsstatus.json file.  Which is in turn, picked up and displayed on the EOSS Tracker home page within the GPS Status box.  Messages include, for example, “waiting on device”, “unable to connect…”, etc.. </li>
<li>The in-memory shared object for GPS status (gps status is shared among the various python processes) is also updated more regularly with GPS status. </li>
<li>More robust handling of connections to GPSD, connections will now reset upon error and attempt a reconnect, etc..  Particularly relevant for off-host GPSD solutions. </li>
<li>More robust handling of various GPS conditions (reported by GPSD) about the GPS device’s status.  </li>
<li>Test mode where gpspoller.py can be run in a standalone fashion to debug GPSD/GPS connection issues. </li>
</ol> 
<li>The kill_session.sh script will delete the /eosstracker/www/daemonstatus.json status file after stopping the processes.  This should [hopefully] prevent the issue where that status file indicates that processes are running when in fact they are not.  For example, in a power outage while the processes are running, upon system reboot, one can hit the “Stop/Abort” button on the EOSS Tracker home page which will run the kill_session.sh script which will NOW delete that daemonstatus.json file.</li>
<li> Updates to the direwolf and other sub-processes so they no longer block waiting on a 3D fix before starting the sub-process.  </li>
</ol>

### CODE REVIEW

File: bin/connectors.py [hunk 1]
```diff
@@ -1611,21 +1620,23 @@ def getPosition(self)->dict:
                             gpsposition["bearing"] = float(position["bearing"])
                             gpsposition["speed_mph"] = float(position["speed_mph"])
                             gpsposition["isvalid"] = True
+                            gpsposition["source"] = "gps"
 
                             # sanity check
                             if gpsposition["latitude"] == 0 or gpsposition["longitude"] == 0:
```
Comment (BradMears) [inline bin/connectors.py:1626]: Is comparing to zero a valid sanity check? Yeah, I know we're not going to have any flights go to the equator or the prime meridian. What I'm really asking is whether the device actually has a "bad data" value vs assuming that zero means the data is bad.

File: bin/connectors.py [hunk 2]
```diff
@@ -1611,21 +1620,23 @@ def getPosition(self)->dict:
...
                                 gpsposition["isvalid"] = False
 
                 if nofix == True:
 
+
                     # we're still waiting on the GPS to obtain a fix so we wait this long
                     seconds = round((1.2) ** trycount) if trycount < 22 else (1.2) ** 22
```
Comment (BradMears) [inline bin/connectors.py]: Three questions:
1) Since the outer loop that starts at 1601 tests for trycount < 2, is testing for trycount < 22 here necessary/useful?
2) What is the purpose of returning 1.2 ** 22 (i.e. 55.2061438912436) in the else?  i.e. What is the significance of 22?
3) If trycount << 22, then seconds will have data type int. If trycount >= 22 and we go through the else, the data type of seconds will be float. Will differing return types cause problems in the calling code? Perhaps the else should have a call to round()?

File: bin/gpspoller.py [hunk 3]
```diff
@@ -206,424 +208,732 @@ def _saveGPSStatus(self, gpsstatus):
...
+                break
+
+            # throttle the loop speed
+            if loopcounter < 10:
+                # Wait short amount of time so the loop isn't spinning at warp speed
+                self.stopevent.wait(short_delay)
+
+            else:
+                # Wait a little longer if we've already been looping for a while
+                self.stopevent.wait(long_delay)
```
Comment (BradMears) [inline bin/gpspoller.py:392]: Good idea to lengthen the delay

File: bin/gpspoller.py [hunk 4]
```diff
@@ -206,424 +208,732 @@ def _saveGPSStatus(self, gpsstatus):
...
+
+
+    ##################################################
+    # connect to GPSD
+    ##################################################
+    def connectToGPSD(self, retrydelay: int = 5, retrylimit: int = 0)->gps:
+
+        # initialization
+        loopcounter = 0
+        retrylimit = retrylimit if retrylimit > 0 else 2178310
```
Comment (BradMears) [inline bin/gpspoller.py:433]: I give up - what is the significance of 2178310? It is almost equal to 2 ** 21 but not quite. A Google search for it turns up nothing.  Enlighten me please

Comment (BradMears) [reply bin/gpspoller.py:433]: Aha! Had I read few lines further I would have seen your array of Fibbonacci numbers and figured it out.

**Review by BradMears — COMMENTED** [review]
Jeff D, Wasn't sure what you wanted when you assigned this issue to me so I just did the same sort of review I do at work.

**Comment by Deatojef** [comment]
This portion of the code is executed when there is a 3D fix from the GPS.  So there should be a reasonable assumption that we have a valid GPS position at this point.  Part of my paranoia here is in the past we’ve had some APRS beacons that would transmit their GPS coords as 0.000,0.000.  So just extending that paranoia into double checking the lat/lon here.  Honestly, we shouldn’t ever have location exactly on 0,0…right?  And if we do, then what are we doing in the Atlantic ocean off the west coast of Africa???Hehe,-JeffSent from my iPadOn Jul 25, 2024, at 3:28 PM, Brad Mears ***@***.***> wrote:﻿
@BradMears commented on this pull request.



In bin/connectors.py:
> @@ -1611,21 +1620,23 @@ def getPosition(self)->dict:
                             gpsposition["bearing"] = float(position["bearing"])
                             gpsposition["speed_mph"] = float(position["speed_mph"])
                             gpsposition["isvalid"] = True
+                            gpsposition["source"] = "gps"
 
                             # sanity check
                             if gpsposition["latitude"] == 0 or gpsposition["longitude"] == 0:

Is comparing to zero a valid sanity check? Yeah, I know we're not going to have any flights go to the equator or the prime meridian. What I'm really asking is whether the device actually has a "bad data" value vs assuming that zero means the data is bad.

File: bin/connectors.py (hunk 1 above)
Comment (Deatojef) [reply bin/connectors.py:1626]: This portion of the code is executed when there is a 3D fix from the GPS.  So there should be a reasonable assumption that we have a valid GPS position at this point.  Part of my paranoia here is in the past we’ve had some APRS beacons that would transmit their GPS coords as 0.000,0.000.  So just extending that paranoia into double checking the lat/lon here.  Honestly, we shouldn’t ever have location exactly on 0,0…right?  And if we do, then what are we doing in the Atlantic ocean off the west coast of Africa???  Hehe

I don't think checking for 0's in this case is impactful to any real locations being used.

-JD

File: bin/connectors.py (hunk 2 above)
Comment (Deatojef) [reply bin/connectors.py]: That ```trycount < 22``` was a hold over from having the connectors (and subprocesses.py) do more “waiting” on the GPS to reach a 3D fix.  Turns out that’s not needed as the subprocess that we’re trying to run can deal with that (ex. Aprsc, direwolf, etc.).  For the connectors.py code, the only legitimate reason for waiting on a 3D fix (i.e. the trycount < 22 business) was for connections to APRS-IS when the brick is set to beacon to the internet.  In that case, we can’t beacon without knowing our locaton…thus the original need to monitor ```trycount``` when the number of times through that while loop became large.  

Long story short, there isn’t a need for that ```trycount < 22``` business anymore.  Good catch!!

File: bin/gpspoller.py (hunk 4 above)
Comment (Deatojef) [reply bin/gpspoller.py:433]: LOL!  Yeah, just needed a way to skip reporting a failed connection attempt to GPSD.  It will retry connection attempts every 5 seconds (by default), but we don’t want to spit out a log message about that every time through the loop.  So the Fibonacci series seemed like a good way to progressively increase the time between log messages as the retry counts get high.  That way we don’t fill up logs with useless retry failure messages every 5 seconds!!

**Comment by TheKoola** [comment]
Thanks @Deatojef -- merged and deleting the `gpsunblock` branch.
